src/MaxText/layers/moe.py

```diff
@@ -747,9 +747,10 @@ def gmm(inputs, kernel, group_sizes, expert_assignments):
       # pad length is the 1st dimension of tiling size in gmm call
       if inputs.shape[0] != expert_assignments.shape[0]:
         raise ValueError("The number of input tokens must match the number of expert" " assignments!")
+      padding_amount = 0
       if hs_shape[0] % pad_length:
-        pad_length = pad_length - hs_shape[0] % pad_length
-        inputs = jax.lax.pad(inputs.astype(jnp.float32), 0.0, [(0, pad_length, 0), (0, 0, 0)])
+        padding_amount = pad_length - hs_shape[0] % pad_length
+        inputs = jax.lax.pad(inputs.astype(jnp.float32), 0.0, [(0, padding_amount, 0), (0, 0, 0)])
 
       inputs = inputs.astype(self.dtype)
       kernel = kernel.astype(self.dtype)
```

Comment:
is there a reason for padding to happen in float32 then casing to self.dtype?

if not, we can do in `self.dtype` with something like this:

```
inputs = inputs.astype(self.dtype)
kernel = kernel.astype(self.dtype)

if hs_shape[0] % pad_length:
    padding_amount = pad_length - hs_shape[0] % pad_length
    zero = jnp.array(0, dtype=inputs.dtype)
    inputs = jax.lax.pad(inputs, zero, [(0, padding_amount, 0), (0, 0, 0)])
```

Reply:
Good catch! Updated to `inputs = jax.lax.pad(inputs, jnp.array(0.0, dtype=inputs.dtype), [(0, padding_amount, 0), (0, 0, 0)])` and tested end-to-end as well.